### PR TITLE
ceph-daemon: Enable "deploy" command  to create a prometheus server instance

### PR DESCRIPTION
**Pull request by pcuzner**
Updates to allow the "ceph-deamon deploy" command to create a Prometheus server container on the local machine. Deploy requires a mgr-list parameter to be passed which is used to configure the initial scrape configuration of Ceph related metrics, from the mgr/prometheus exporter.

Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

### CODE REVIEW

File: src/ceph-daemon/ceph-daemon [hunk 1]
```diff
@@ -256,7 +299,10 @@ def makedirs(dir, uid, gid, mode):
 
 def get_data_dir(fsid, t, n):
     # type: (str, str, Union[int, str]) -> str
-    return os.path.join(args.data_dir, fsid, '%s.%s' % (t, n))
+    if n:
+        return os.path.join(args.data_dir, fsid, '%s.%s' % (t, n))
+    else:
+        return os.path.join(args.data_dir, fsid, '%s' % (t))
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: we could name the prometheus instances like ``prometheus.cxwads`` like we do the mgr etc and avoid this weirdness...

File: src/ceph-daemon/ceph-daemon [hunk 2]
```diff
@@ -411,19 +457,26 @@ def get_legacy_daemon_fsid(cluster, daemon_type, daemon_id, legacy_dir=None):
...
+        r += metadata['args']
     return r
 
 def create_daemon_dirs(fsid, daemon_type, daemon_id, uid, gid,
                        config=None, keyring=None):
     # type: (str, str, Union[int, str], int, int, str, str) ->  None
-    data_dir = make_data_dir(fsid, daemon_type, daemon_id)
-    make_log_dir(fsid)
+    data_dir = make_data_dir(fsid, daemon_type, daemon_id, uid, gid)
+    make_log_dir(fsid, uid, gid)
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: this hunk looks like a bug independent of this pr?

File: src/ceph-daemon/ceph-daemon [hunk 3]
```diff
@@ -834,6 +939,9 @@ def deploy_crash(fsid, uid, gid, config, keyring):
 
 def get_unit_file(fsid, uid, gid):
     # type: (str, int, int) -> str
+    #
+    # FIXME - should monitoring containers be included in the ceph.target group?
+    #
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: IMO yes

File: src/ceph-daemon/ceph-daemon [hunk 4]
```diff
@@ -1265,13 +1377,18 @@ def command_bootstrap():
...
             c_out = cli(['-h'])
             if 'dashboard' in c_out:
                 break
-            logger.info('Dashboard not yet available, waiting...')
+            logger.info('Dashboard not yet available, waiting...({})'.format(limit))
             time.sleep(1)
+            limit -= 1
+            if limit == 0:
+                raise Error("Timeout waiting for dashboard to be available")
+
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: put this in a separate patch?

File: src/ceph-daemon/ceph-daemon [hunk 5]
```diff
@@ -1644,7 +1799,12 @@ def command_adopt():
 
 def command_rm_daemon():
     # type: () -> None
-    (daemon_type, daemon_id) = args.name.split('.', 1)
+    if '.' in args.name:
+        (daemon_type, daemon_id) = args.name.split('.', 1)
+    else:
+        daemon_type = args.name
+        daemon_id = None
+
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: sigh this is everywhere.. makes me liek teh idea of sticking to $type.$id even when we only expect one of them

File: src/ceph-daemon/ceph-daemon [hunk 6]
```diff
@@ -2007,6 +2168,10 @@ def _get_parser():
     parser_deploy.add_argument(
         '--config-and-keyrings',
         help='JSON file with config and keyrings for the daemon and crash agent')
+    parser_deploy.add_argument(
+        '--mgr-list',
+        help='Comma separated list of ceph mgr IP address or names e.g. mgr-1,mgr2'
+    )
```
Comment (liewegas) [inline src/ceph-daemon/ceph-daemon]: i wonder if instead of adding this arg we change the --config-and-keyrings above to --config-json or something and bundle everything in there.  becuase for instance the grafana is going to take all kinds of random info from mgr/ssh.  And we could move some of the prometheus config that is hardcoded here into mgr/ssh too.

File: src/ceph-daemon/ceph-daemon (hunk 3 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: ACK

File: src/ceph-daemon/ceph-daemon (hunk 1 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: what about accepting prometheus as the name, but in deploy and remove, if the id is None, set it to 0 - then everything else stays the same?

File: src/ceph-daemon/ceph-daemon (hunk 5 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: see prior comment about defaulting to 0 for the id if not present

File: src/ceph-daemon/ceph-daemon (hunk 4 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: Point taken - but if we agree we need it, why not keep in and save a further PR and review process?

File: src/ceph-daemon/ceph-daemon (hunk 2 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: I had to fix it for this PR - since the uid/gid is set by the default extract_uid_gid which expects the image to be ceph. For monitoring it's not - so it had to be fixed.

File: src/ceph-daemon/ceph-daemon (hunk 6 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: Not sure about repurposing config-and-keyrings - the function returns a tuple of 3 and re very ceph specific. why not replace mgr-list and go for a generic --config-json this could be in the install payload and also the parameter for the reconfig payload?

File: src/ceph-daemon/ceph-daemon (hunk 4 above)
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: @pcuzner please see PR #32049. Looks like we are attempting to solve the same issue, but I'm aok with either approach ...

File: src/ceph-daemon/ceph-daemon (hunk 6 above)
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: sounds good to me

File: src/ceph-daemon/ceph-daemon (hunk 4 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: @mgfritch I'll back my change out, in favour of your approach.

File: src/ceph-daemon/ceph-daemon (hunk 1 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: @liewegas updated to use <service>.<hostname> so the dir for the monitoring components will look more like mon/mgr dirs

File: src/ceph-daemon/ceph-daemon (hunk 5 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: see above - latest commit uses hostname as the daemon_id

File: src/ceph-daemon/ceph-daemon (hunk 6 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: config-json added

File: src/ceph-daemon/ceph-daemon (hunk 5 above)
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: I would just have the user pass ``--name prometheus.foo`` instead of silently add it--less surprising.  (And also mgr/ssh can then use whatever naming convention it likes)

Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: ACK

File: src/ceph-daemon/ceph-daemon [hunk 7]
```diff
@@ -436,6 +503,81 @@ def create_daemon_dirs(fsid, daemon_type, daemon_id, uid, gid,
...
+        with open(os.path.join(data_dir_root,'etc/prometheus/prometheus.yml'), 'w') as f:
+            os.fchown(f.fileno(), uid, gid)
+            os.fchmod(f.fileno(), 0o600)
+            f.write(yaml.safe_dump(prometheus_yml, explicit_start=True, indent=2))
+
+def get_parm(option):
+    """Read requested arg, and return as a dict"""
+
+    if not option:
+        return None
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: :-1: for silently returning `None`. Can we raise an exception here?

File: src/ceph-daemon/ceph-daemon [hunk 8]
```diff
@@ -436,6 +503,81 @@ def create_daemon_dirs(fsid, daemon_type, daemon_id, uid, gid,
...
+
+    if option == '-':
+        try:
+            j = injected_stdin
+        except NameError:
+            j = sys.stdin.read()
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: Yet another copy of the

```python
try:
            j = injected_stdin
        except NameError:
            j = sys.stdin.read()
```

ideom

File: src/ceph-daemon/ceph-daemon [hunk 9]
```diff
@@ -436,6 +503,81 @@ def create_daemon_dirs(fsid, daemon_type, daemon_id, uid, gid,
...
+    else:
+        # inline json string
+        if option[0] == '{' and option[-1] == '}':
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: I can't see any code in this PR that make use inline JSON for arguments. Can we remove this?

File: src/ceph-daemon/ceph-daemon [hunk 10]
```diff
@@ -1305,24 +1477,66 @@ def command_bootstrap():
...
     # type: () -> None
     (daemon_type, daemon_id) = args.name.split('.', 1)
-    if daemon_type not in ['mon', 'mgr', 'mds', 'osd', 'rgw', 'rbd-mirror']:
+    
+    crash_keyring = None
+    
+    supported_daemons = Ceph.daemons.copy()
+    supported_daemons.extend(Monitoring.components)
+
+    if daemon_type not in supported_daemons:
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]:
```suggestion
    if daemon_type not in Ceph.daemons and daemon_type not in Monitoring.components:
```

File: src/ceph-daemon/ceph-daemon [hunk 11]
```diff
@@ -1305,24 +1477,66 @@ def command_bootstrap():
...
+            elif args.mon_addrv:
+                config += '[mon.%s]\n\tpublic_addrv = %s\n' % (daemon_id,
+                                                            args.mon_addrv)
+            elif args.mon_network:
+                config += '[mon.%s]\n\tpublic_network = %s\n' % (daemon_id,
+                                                                args.mon_network)
+            else:
+                raise Error('must specify --mon-ip or --mon-network')
+    
+    if daemon_type in Ceph.daemons:
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: This `if` seems superfluous to me. Can we simply remove it?

**Comment by sebastian-philipp** [comment]
Seing lots of type errors:

```
ceph-daemon:361: error: Argument 2 to "make_data_dir_base" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:361: error: Argument 3 to "make_data_dir_base" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:363: error: Argument 2 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:363: error: Argument 3 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:371: error: Argument 2 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:371: error: Argument 3 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:387: error: Argument 2 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:387: error: Argument 3 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:411: error: Argument 2 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:411: error: Argument 3 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:496: error: Need type comment for 'r' (hint: "r = ...  # type: List[<type>]")
ceph-daemon:508: error: Value of type "object" is not indexable
ceph-daemon:539: error: Need type comment for 'prometheus_yml' (hint: "prometheus_yml = ...  # type: Dict[<type>, <type>]")
ceph-daemon:543: error: "str" has no attribute "get"
ceph-daemon:547: error: Value of type "object" is not indexable
ceph-daemon:548: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
ceph-daemon:692: error: Argument 3 to "get_data_dir" has incompatible type "Union[int, str, None]"; expected "Union[int, str]"
ceph-daemon:716: error: Incompatible types in assignment (expression has type "None", variable has type "str")
ceph-daemon:733: error: Type signature has too few arguments
ceph-daemon:1081: error: Need type comment for 'entrypoint' (hint: "entrypoint = ...  # type: List[<type>]")
ceph-daemon:1512: error: Need type comment for 'monitoring_args' (hint: "monitoring_args = ...  # type: List[<type>]")
ceph-daemon:1531: error: "object" has no attribute "get"
ceph-daemon:1533: error: "object" has no attribute "get"
Found 23 errors in 1 file (checked 1 source file)
```

**Comment by liewegas** [comment]
rebase?

File: src/ceph-daemon/ceph-daemon (hunk 2 above)
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: ha! looks as if we are attempting to solve the same thing: https://github.com/ceph/ceph/pull/32137/commits/5111a1145946a7b6005a68ec7e9bf19b88f60524#diff-0d7e098eacbe52f3a7f86d925dfeff1fR450

@pcuzner I'll drop my change in favor of yours, but instead positional args can we specify the keywords?
```suggestion
    data_dir = make_data_dir(fsid, daemon_type, daemon_id, uid=uid, gid=gid)
    make_log_dir(fsid, uid=uid, gid=gid)
```

Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: err, actually sorry, guess PR #32137 was just merged. But :+1: for adding uid/gid to `make_log_dir`

File: src/ceph-daemon/ceph-daemon (hunk 8 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: @sebastian-philipp Yes - it's another copy - that's why I've tried to make get_parm generic - so we can hook up the other functions to get parm and do this in one place.

File: src/ceph-daemon/ceph-daemon (hunk 9 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: see comment above about get_parm becoming the generic json parser

File: src/ceph-daemon/ceph-daemon (hunk 11 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: it's not superfluous. We need a branch in the code to cater for the way we determine the uid/gid to set. For ceph daemons, we look at /etc/ceph - for prometheus et al it's a different file.

File: src/ceph-daemon/ceph-daemon (hunk 7 above)
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: We return none, and in the caller decide whether none is a problem or not. in this case it is and we raise an exception.

File: src/ceph-daemon/ceph-daemon [hunk 12]
```diff
@@ -53,6 +53,8 @@ import subprocess
 import sys
 import tempfile
 import time
+import yaml
```
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: grafana/prom and alertmanager all have yaml config files.

File: src/ceph-daemon/ceph-daemon [hunk 13]
```diff
@@ -69,6 +71,64 @@ container_path = None
...
+        }
+    }
+
+
+def port_in_use(port_num):
+    """Detect whether a port is in use on the local machine"""
+
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.connect(("127.0.0.1",port_num))
```
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: will update

File: src/ceph-daemon/ceph-daemon (hunk 12 above)
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: Not sure that your asking to be done here. I use yaml.safe_dump to create the files for prom (and grafana, alertmanager later) and it also provides a simpler update process when adding hosts, updating scrape configs etc in a reliable way. 

I could catch the yaml import issue, and make requests for monitoring basically become a NO-OP.

Appreciate some guidance here on how to resolve
@liewegas @sebastian-philipp

[the remark this answers is not in the document]
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: I'm not sure how to proceed.  We want this to work with a minimal base OS (base python install).  Are we just dumping (but not parsing) yaml?  Maybe there is an open-coded yaml dumper we can include or something. Bleh..

[the remark this answers is not in the document]
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: can we handle the configs as strings and substitute in values as needed?

[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: mgr/ssh is going to send config json for deployment and reconfig operations, so a simple string substitution may not be flexible or robust enough. 
If we can't use yaml in ceph-daemon, we could push all handling of config for the monitoring daemons up into mgr/ssh (assuming yaml is available within and ceph-daemon - it is for py2, but not py3 AFAICT). ceph-deamon then simply receives the yaml string to commit to file.This implies that mgr/ssh needs to maintain & persist config state for all monitoring component files - which may help anyway, and lead to the UI exposing/managing the config content?

???

File: src/ceph-daemon/ceph-daemon (hunk 13 above)
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: done

File: src/ceph-daemon/ceph-daemon (hunk 12 above)
[the remark this answers is not in the document]
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: What does ceph-daemon really need to do to configure it?  It seems easier to just pass the config file(s) in as strings.  maybe even a json dict of filename -> file content.

**Comment by sebastian-philipp** [comment]
```
==================================== ERRORS ====================================
__________________ ERROR collecting tests/test_ceph_daemon.py __________________
ImportError while importing test module '/home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-daemon/tests/test_ceph_daemon.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_ceph_daemon.py:7: in <module>
    cd = imp.load_source("ceph-daemon", "ceph-daemon")
ceph-daemon:57: in <module>
    import yaml
E   ImportError: No module named yaml
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.15 seconds ============================
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-daemon/.tox/py27/bin/pytest (exited with code 2)
mypy create: /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-daemon/.tox/mypy
mypy installdeps: mypy
WARNING: Discarding $PYTHONPATH from environment, to override specify PYTHONPATH in 'passenv' in your configuration.
mypy installed: -f /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-daemon/wheelhouse,mypy==0.750,mypy-extensions==0.4.3,typed-ast==1.4.0,typing-extensions==3.7.4.1
mypy run-test-pre: PYTHONHASHSEED='1376554495'
mypy run-test: commands[0] | mypy ceph-daemon
ceph-daemon:365: error: Argument 2 to "make_data_dir_base" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:365: error: Argument 3 to "make_data_dir_base" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:367: error: Argument 2 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:367: error: Argument 3 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:375: error: Argument 2 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:375: error: Argument 3 to "makedirs" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:395: error: Argument 2 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:395: error: Argument 3 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:420: error: Argument 2 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:420: error: Argument 3 to "chown" has incompatible type "Optional[int]"; expected "int"
ceph-daemon:503: error: Need type annotation for 'r' (hint: "r: List[<type>] = ...")
ceph-daemon:515: error: Value of type "object" is not indexable
ceph-daemon:546: error: Need type annotation for 'prometheus_yml' (hint: "prometheus_yml: Dict[<type>, <type>] = ...")
ceph-daemon:550: error: "str" has no attribute "get"
ceph-daemon:554: error: Value of type "object" is not indexable
ceph-daemon:555: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
ceph-daemon:699: error: Argument 3 to "get_data_dir" has incompatible type "Union[int, str, None]"; expected "Union[int, str]"
ceph-daemon:723: error: Incompatible types in assignment (expression has type "None", variable has type "str")
ceph-daemon:740: error: Type signature has too few arguments
ceph-daemon:1088: error: Need type annotation for 'entrypoint' (hint: "entrypoint: List[<type>] = ...")
ceph-daemon:1531: error: Need type annotation for 'monitoring_args' (hint: "monitoring_args: List[<type>] = ...")
ceph-daemon:1550: error: "object" has no attribute "get"
ceph-daemon:1552: error: "object" has no attribute "get"
ceph-daemon:2008: error: Name 'check_time_sync' already defined on line 1976
ceph-daemon:2018: error: Name 'command_check_host' already defined on line 1986
Found 25 errors in 1 file (checked 1 source file)
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-daemon/.tox/mypy/bin/mypy ceph-daemon (exited with code 1)
```

File: src/ceph-daemon/ceph-daemon (hunk 11 above)
Comment (sebastian-philipp) [reply src/ceph-daemon/ceph-daemon]: Yeah, I meant: If we just remove this specific line (1507) it should still work?

File: src/ceph-daemon/ceph-daemon [hunk 14]
```diff
@@ -436,6 +507,81 @@ def create_daemon_dirs(fsid, daemon_type, daemon_id, uid, gid,
...
+            else:
+                raise Error("config-json must either specify a valid prometheus.yml, or a mgrs list")
+
+        # create configs
+        with open(os.path.join(data_dir_root,'etc/prometheus/prometheus.yml'), 'w') as f:
+            os.fchown(f.fileno(), uid, gid)
+            os.fchmod(f.fileno(), 0o600)
+            f.write(yaml.safe_dump(prometheus_yml, explicit_start=True, indent=2))
+
+def get_parm(option):
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]:
```suggestion
def get_parm(option):
    # type: (str) -> Optional[Any]
```

File: src/ceph-daemon/ceph-daemon (hunk 12 above)
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: @liewegas yep - will revert to that approach

**Comment by pcuzner** [comment]
Closing PR - will open another PR against the renamed code in cephadm dir
